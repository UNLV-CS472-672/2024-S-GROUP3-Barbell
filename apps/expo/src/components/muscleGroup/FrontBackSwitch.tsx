import React, { useEffect, useRef, useState } from 'react'
import {
  Animated,
  Dimensions,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native'

// screen dimensions
// Scaled based on screenWidth and screenHeight
const { width: screenWidth, height: screenHeight } = Dimensions.get('window')

// Custom switch component show at bottom - With Animation
// Call as follows:
// <onValueChange
//     label="..."
//     value={useState()}
//     onValueChange={() => set(!On)}
// />
export default function FrontBackSwitch ({
  value,
  onValueChange,
  label,
}: {
  value: boolean
  onValueChange: () => void
  label: string
})  {
  // Use useRef to persist the animated value without reinitializing it on every render
  const animation = useRef(new Animated.Value(value ? 1 : 0)).current

  useEffect(() => {
    // animate the switch when the value prop changes
    Animated.timing(animation, {
      // Animate to 1 if enabled, else to 0
      toValue: value ? 1 : 0,
      duration: 150,
      // since we're animating a non-native property (marginLeft keke)
      useNativeDriver: false,
    }).start()
  }, [value, animation])

  // interpolate the animated value to calculate marginLeft
  // Use screen width for compat
  const circleTransform = animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, screenWidth * 0.35],
  })

  const backgroundColor = animation.interpolate({
    inputRange: [0, 1],
    // Simple left to right
    // No effects
    outputRange: ['#797979', '#797979'],
  })

  return (
    <View style={styles.toggleContainer}>
      <TouchableOpacity onPress={onValueChange} activeOpacity={1}>


        <Animated.View style={[styles.switch, { backgroundColor }]}>
          {/* Animated the marginLeft of the circle */}
          <Animated.View
            style={[styles.circle, { marginLeft: circleTransform }]}
          />
        </Animated.View>

        <Text style={{fontSize: screenWidth * 0.04, textAlign: 'center',
          margin: 10, position: 'absolute', color: '#CACACA',
          width: screenWidth * 0.7, height: screenWidth * 0.1,
          top: 0, left: 0-(screenWidth*0.4)/2  }}>FRONT
        </Text>

        <Text style={{fontSize: screenWidth * 0.04, textAlign: 'center',
          margin: 10, position: 'absolute', color: '#CACACA',
          width: screenWidth * 0.7, height: screenWidth * 0.1,
          top: 0, left: (screenWidth*0.3)/2  }}>BACK
        </Text>

      </TouchableOpacity>
      <Text style={styles.label}>{label}</Text>
    </View>
  )
}


// For rectangle shape
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#1C1B1B',
  },
  toggleContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: screenWidth * 0.015,
  },
  switch: {
    width: screenWidth * 0.7,
    height: screenWidth * 0.1,
    // Adjust shape to rectangle with round edges
    borderRadius: screenWidth * 0.02,
    justifyContent: 'center',
    // Handle thin line on the right
    marginRight: screenWidth * 0.012,
  },
  circle: {
    width: screenWidth * 0.35,
    height: screenWidth * 0.1,
    borderRadius: screenWidth * 0.02,
    backgroundColor: '#48476D',
  },
  label: {
    // Overlap
    fontSize: screenWidth * 0.04,
    color: '#CACACA',
  }
})
