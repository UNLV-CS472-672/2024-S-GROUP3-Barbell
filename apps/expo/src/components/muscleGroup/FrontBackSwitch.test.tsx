import { render, screen } from '@testing-library/react-native'

import FrontBackSwitch from "~/components/muscleGroup/FrontBackSwitch";
import {useState} from "react";

test('FrontBackSwitch', () => {
  const [tmp, setTmp] = useState(false)
  const tmpCh = () => setTmp((previousState) => !previousState)

  render(<FrontBackSwitch
    onValueChange={tmpCh}
    value={tmp}
    label={""}
  />)
  expect(screen.toJSON()).toMatchSnapshot()
})
