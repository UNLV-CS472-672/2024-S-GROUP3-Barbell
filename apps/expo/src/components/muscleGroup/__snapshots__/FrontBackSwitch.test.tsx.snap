// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontBackSwitch should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "marginBottom": 11.25,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(121, 121, 121, 1)",
          "borderRadius": 15,
          "height": 75,
          "justifyContent": "center",
          "marginRight": 9,
          "width": 525,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#48476D",
            "borderRadius": 15,
            "height": 75,
            "marginLeft": 0,
            "width": 262.5,
          }
        }
      />
    </View>
    <Text
      style={
        {
          "color": "#CACACA",
          "fontSize": 30,
          "height": 75,
          "left": -150,
          "margin": 10,
          "position": "absolute",
          "textAlign": "center",
          "top": 0,
          "width": 525,
        }
      }
    >
      FRONT
    </Text>
    <Text
      style={
        {
          "color": "#CACACA",
          "fontSize": 30,
          "height": 75,
          "left": 112.5,
          "margin": 10,
          "position": "absolute",
          "textAlign": "center",
          "top": 0,
          "width": 525,
        }
      }
    >
      BACK
    </Text>
  </View>
  <Text
    style={
      {
        "color": "#CACACA",
        "fontSize": 30,
      }
    }
  >
    test
  </Text>
</View>
`;
